### DISCUSSION TITLE
Exploring the Plotting of Pair Correlation Function for Gaussian Potentials using DeepSets Neural Networks in Continuous Space

### DISCUSSION THREAD

**Discussion by MrAAafa99** [netket/netket#1482] [Q&A] [answered]
Hello,
I am trying to regenerate pair correlation functions for Gaussian potentials in continuous space using the Deep Sets neural network that the results of this article (https://arxiv.org/abs/2112.11957)
here is the code that I wrote( in this code I have 10 particles and density=1.0):

```python
import netket as nk
import jax.numpy as jnp
import matplotlib.pyplot as plt
N = 10
L = N / 1.0
hilb = nk.hilbert.Particle(N=N, L=L, pbc=True)
def minimum_distance(x, sdim):
    """Computes distances between particles using minimum image convention"""
    n_particles = x.shape[0] // sdim
    x = x.reshape(-1, sdim)
    distances = (-x[jnp.newaxis, :, :] + x[:, jnp.newaxis, :])[
        jnp.triu_indices(n_particles, 1)
    ]
    distances = jnp.remainder(distances + L / 2.0, L) - L / 2.0
    return jnp.linalg.norm(distances, axis=1)

sigma = 1 / (2 * jnp.sqrt(2 * jnp.log(2)))
epsilon = 2

def potential(x, sdim):
    dis = minimum_distance(x, sdim)
    return jnp.sum(jnp.exp(-0.5 * (dis / sigma)**2))

ekin = nk.operator.KineticEnergy(hilb, mass=1.0)
pot = nk.operator.PotentialEnergy(hilb, lambda x: potential(x, 1))
ha = ekin + pot

sab = nk.sampler.MetropolisGaussian(hilbert=hilb, sigma=0.05, n_chains=16, n_sweeps=32)
model = nk.models.DeepSetRelDistance( hilbert=hilb, layers_phi=2, layers_rho=3, features_phi=(16, 16), features_rho=(16, 16, 1), cusp_exponent=None)
vs = nk.vqs.MCState(sab, model, n_samples=4096, n_discard_per_chain=128)

op = nk.optimizer.Sgd(0.001)
sr = nk.optimizer.SR(diag_shift=0.01)
gs = nk.VMC(ha, op, sab, variational_state=vs, preconditioner=sr)
gs.run(n_iter=10, out="Gaussian_Potential_10_1d")
```

and for calculation and plotting pair correlation function, I wrote this code:

```python
import numpy as np
hist_bins = 50
hist_range = (0, L/2.0)

all_histograms = []
for sample in vs.samples:
    d = minimum_distance(sample, 1)
    hist, _ = np.histogram(d, bins=hist_bins, range=hist_range, density=True)
    all_histograms.append(hist)

 ##Computing average histogram##
average_hist = np.mean(all_histograms, axis=0)
bin_edges = np.linspace(hist_range[0], hist_range[1], hist_bins+1)

 ##Plotting histograms for all samples##
for hist in all_histograms:
    plt.plot(bin_edges[:-1], hist, alpha=0.3)

 ##Plotting the average histogram##
plt.plot(bin_edges[:-1], average_hist, 'r', linewidth=2)

plt.xlabel('Distance')
plt.ylabel('Probability Density')
plt.title('Histogram of Distances')
plt.show()
```

To calculate and plot the pair correlation function from transformed samples, I used a histogram to plot the distribution of a single sample (using "np.histogram"). Then, I repeated this process for all samples, plotted histograms for each of them, and calculated the average of all histograms. Finally, I plotted the pair correlation function.
When I plot the pair correlation function for a Gaussian potential, it will resemble the shape shown below(Note: the Red line is the averaging over all plots):

![Gausiann potential](https://github.com/netket/netket/assets/54255011/c52c85cf-c854-40ce-afb9-c007d56dbbf2)

but the real result in the article is the figure below:

![Screenshot (753)](https://github.com/netket/netket/assets/54255011/9ea4c094-e59b-4339-be7a-f8de0accf557)

I have tried the run this code for different potentials that have Kato’s Cusp exponent. It involves plotting the potential shape and obtaining the corresponding results. The process is as follows, for example, for Aziz Potential  the figure of the pair correlation function looks like this :

![Aziz potential](https://github.com/netket/netket/assets/54255011/cf343035-43ce-4985-8c51-9d9fdf2e926b)

This figure visually reproduces the shape of the article's figure titled "Neural-Network Quantum States for Periodic Systems in Continuous Space".

I'm not sure why I'm not getting reasonable results for Gaussian potential. I would like to understand the underlying cause and find a solution to obtain meaningful outcomes. Your assistance in resolving this issue would be greatly appreciated.

**Comment by PhilipVinc** [comment]
@gpescia is an author of the paper and might know..

**Comment by PhilipVinc** [comment]
Though with such a tiny learning rate and short optimisation, I'm not surprised you get... nothing meaningful?

**Comment by gpescia** [comment]
First of all I agree with @PhilipVinc that you cannot expect reasonable results for only 10 optimization steps. Would you mind sharing your training curve? 
To have an idea if you have converged or not, you can check the variance of the energy which should go towards zero. You can also compare the energy per particle to the values you find in the paper you cited. The exact numbers will not match but the order of magnitude will be the same. 
For faster convergence I would also suggest to decrease the diagonal shift of the SR and increase the learning rate. 

The way you compute the radial distribution looks not fully correct. I think you should not be using `density=True` but do the normalization yourself: divide the resulting unnormalized histograms by the number density of your system. I also suspect that you need to flatten the array of distances `d` to 1D before computing the histogram.

Does that help?

**Comment by MrAAafa99** [comment]
@PhilipVinc and @gpescia, thank you for your response. I optimized the algorithm up to 1000 iterations, and I obtained the following curve-fitting graph for the 1000 steps. It looks like this.

![download](https://github.com/netket/netket/assets/54255011/b52c077f-fe91-4ffd-aa5e-937baacb5673)



I made sure that my energy has converged. I also rewrote the code with the corrections proposed by @gpescia. Here is my code:

       import numpy as np
       hist_bins = 50
       hist_range = (0, L/2.0)


       all_histograms = []
       for sample in vs.samples:
           d = np.ravel(minimum_distance(sample, 1))
           hist, _ = np.histogram(d, bins=hist_bins, range=hist_range)
           all_histograms.append(hist)

        ## Computing average histogram##
       average_hist = np.mean(all_histograms, axis=0)
       bin_edges = np.linspace(hist_range[0], hist_range[1], hist_bins+1)

       ## Normalize histograms by number density##
       number_density = N / (L ** 1)  # Assuming sdim=1
       average_hist = average_hist.astype(float) / number_density
       for i in range(len(all_histograms)):
           all_histograms[i] = all_histograms[i].astype(float) / number_density

        ## Plotting histograms for all samples##
       for hist in all_histograms:
           plt.plot(bin_edges[:-1], hist, alpha=0.3)

           ## Plotting the average histogram##
       plt.plot(bin_edges[:-1], average_hist, 'r', linewidth=2)

       plt.xlabel('Distance')
       plt.ylabel('Probability Density')
       plt.title('Histogram of Distances')
       plt.ylim(1.5, 4)  # Set the y-axis limits
       plt.show()


And this is the plot that I got.

![Gaussian_1000_1D](https://github.com/netket/netket/assets/54255011/865410a2-faed-42cf-8159-5d581a9360f9)


 It starts from 2 to 2.5  But it must start from less than one and reach one when I normalize the pair correlation function. However, as you can see in the plot, I have a lot of noise.
Thank you for your assistance.

**Reply by gpescia** [reply]
The energy does not look correct, which makes sense since your sigma in the potential is different from the paper. If I remember correctly it should just be 0.5. Then the energy per particle should be somewhere in the ballpark of 1.23. 
Could you send me the samples that you are using to compute the radial distribution function? I want to check if your code to compute it has a bug or if the samples are not correct.

**Reply by MrAAafa99** [reply]
according to the paper for the Gaussian potentials, sigma is equal to 1/sqrt(2) and epsilon=2. 
I have modified these two variables in the above code, and as you can see, my training curve for Gaussian potential, N=20 and D=1 are displayed as follows:

![download](https://github.com/netket/netket/assets/54255011/adda62a1-9594-49b8-ae2e-f4d142d509a3)

The converted energy became 24.878, and the energy per particle became 1.24, which precisely matches the findings in the article.


When plotting the radial distribution function from the above code that I wrote in the previous comment,  for hist_bins = 50, I obtain the following figure:


![GP1D203 bin50](https://github.com/netket/netket/assets/54255011/ec5065bb-e357-4610-9341-e7be6eaf4d1c)

but when I plot it for hist_bin=150, I obtain the following figure:

![Gp1D203_bin150](https://github.com/netket/netket/assets/54255011/048f682a-86f9-481a-804c-7ffb4fa8d279)

It seems logical to me, but when I increase the number of bins, for example, up to 200, it becomes illogical again! and the following figure for hist_bin=200 is look like this:

![GP1D203 bin200](https://github.com/netket/netket/assets/54255011/d1f5a9e4-86a3-4922-8630-4216193c2726)

I have also emailed you the samples I used to calculate the radial distribution function since I'm not sure how to upload the .mpack file here while keeping it as a file.

**Reply by gpescia** [reply]
Ok nice I guess if you match the energy there is a problem with the code for the distribution function. I did not get the samples yet though. Also: Can you save the samples in a .txt file or .npy file or smth similar. Otherwise I'm not sure it is easy to reload them from the .mpack file, since I don't have your model.

**Reply by MrAAafa99** [reply]
I emailed it to you again, and I think it should have arrived by now. However, since it takes a lot of time for reaching convergence to properly for saving it again as .text or.npy, I believe it would be better to provide you with the model in the form of code so that you can reload the data from .mpack file :
`




       import netket as nk
       import jax.numpy as jnp
       import matplotlib.pyplot as plt
       import flax
       import math
       N = 20
       L = N / 1.0

       hilb = nk.hilbert.Particle(
           N=N,
           L=L,
           pbc=True
       )

       def minimum_distance(x, sdim):
           """Computes distances between particles using minimum image convention"""
           n_particles = x.shape[0] // sdim
           x = x.reshape(-1, sdim)

           distances = (-x[jnp.newaxis, :, :] + x[:, jnp.newaxis, :])[
               jnp.triu_indices(n_particles, 1)
           ]
           distances = jnp.remainder(distances + L / 2.0, L) - L / 2.0

           return jnp.linalg.norm(distances, axis=1)

       sigma = 1 / math.sqrt(2)
       epsilon = 2

       def potential(x, sdim):
           dis = minimum_distance(x, sdim)
           return jnp.sum( epsilon * (jnp.exp(-(dis)**2)))

       ekin = nk.operator.KineticEnergy(hilb, mass=1.0)
       pot = nk.operator.PotentialEnergy(hilb, lambda x: potential(x, 1))
       ha = ekin + pot

       sab = nk.sampler.MetropolisGaussian(hilbert=hilb, sigma=0.05, n_chains=16, n_sweeps=32)
       model = nk.models.DeepSetRelDistance(
           hilbert=hilb,
           layers_phi=2,
           layers_rho=3,
           features_phi=(16, 16),
           features_rho=(16, 16, 1),
           cusp_exponent=None
       )
       vs = nk.vqs.MCState(sab, model, n_samples=4096, n_discard_per_chain=128)

       op = nk.optimizer.Sgd(0.001)
       sr = nk.optimizer.SR(diag_shift=0.01)
       gs = nk.VMC(ha, op, sab, variational_state=vs, preconditioner=sr)
       gs.run(n_iter=1000, out="Gaussian_20_1D")
`

**Reply by gpescia** [reply]
I think you have too few samples but modifying your code to this, gives me a reasonable result:
```
N = 20
L = N / 1.0

import numpy as np
hist_bins = 100
hist_range = (0, L/2.0)

samples = np.load('vs_samples.npy').reshape(-1,20)
all_histograms = []
for sample in samples:
    d = np.ravel(minimum_distance(sample, 1))
    hist, _ = np.histogram(d, bins=hist_bins, range=hist_range, density=True)
    all_histograms.append(hist)

## Computing average histogram##
average_hist = np.mean(all_histograms, axis=0)/2.
bin_edges = np.linspace(hist_range[0], hist_range[1], hist_bins+1)
## Normalize histograms by number density##
number_density = 1 / (L ** 1)  # Assuming sdim=1
average_hist = average_hist.astype(float) / number_density

plt.plot(average_hist)
plt.xlabel('Distance')
plt.ylabel('Probability Density')
plt.title('Histogram of Distances')
plt.show()
```

So 1) reshape your samples to the correct shape and 2) an additional factor of 2 when computing the average histogram. 
Btw to make this run faster you can replace the for-loop with a fori_loop primitive of jax or a vmap. 
Does this help?

**Reply by MrAAafa99** [reply]
Yes, the modified code has been very helpful. I have plotted the g(r) graph using your modified code, and I obtained the following graph.

![Probability Density](https://github.com/netket/netket/assets/54255011/c05acdc6-bc06-44ae-8c52-0d021128b0e8)

 It appears to be logical, but as you can see even when the g(r) converges, it still exhibits fluctuations. Does this seem reasonable? Or is there a way to reduce the fluctuations when the g(r) converges to 1?

**Reply by gpescia** [reply]
Can you try to increase the number of samples to reduce the noise?

**Reply by MrAAafa99** [reply]
Yes, I am increasing the number of samples, but it does not improve the fluctuations present in the radial distribution plot. I believe the issue lies with the radial distribution code.
According to the modified code, the radial distribution, which is as follows, 

       N = 20
       L = N / 1.0

       import numpy as np
       hist_bins = 100
       hist_range = (0, L/2.0)

       samples = np.load('vs_samples.npy').reshape(-1,20)
       all_histograms = []
       for sample in samples:
           d = np.ravel(minimum_distance(sample, 1))
           hist, _ = np.histogram(d, bins=hist_bins, range=hist_range, density=True)
           all_histograms.append(hist)

       ## Computing average histogram##
       average_hist = np.mean(all_histograms, axis=0)/2.
       bin_edges = np.linspace(hist_range[0], hist_range[1], hist_bins+1)
       ## Normalize histograms by number density##
       number_density = 1 / (L ** 1)  # Assuming sdim=1
       average_hist = average_hist.astype(float) / number_density

       plt.plot(average_hist)
       plt.xlabel('Distance')
       plt.ylabel('Probability Density')
       plt.title('Histogram of Distances')
       plt.show()

When I plot the radial distribution for the number of samples equal to 8192 instead of 4096, for N=20 and D=1.0 (which means L=N/D=20) and set the"hist_bins" to 100, I obtain the following graph:
![download](https://github.com/netket/netket/assets/54255011/e458d8fd-d119-40cd-8ff1-adfbde139b22)
As can be seen, the fluctuations are noticeable in the graph.
However, since in this example L=20, I believe I should set the"hist_bins" to 20 because our total length is equal to 20. As you can see, I plotted the"hist_bins" as 20, and the radial distribution plot is shown below.

![download (1)](https://github.com/netket/netket/assets/54255011/b9c3ca9c-6b64-4bc0-b084-d180699b9ec5)

As can be observed, the fluctuations in the radial distribution plot have decreased, which is logical since the distance has been reduced.

However, as shown in the figure below, taken from the article (https://arxiv.org/abs/2112.11957) , 

![Screenshot (776)](https://github.com/netket/netket/assets/54255011/0e20b987-e36f-4ebc-9a27-0b5624863cbe)


for a one-dimensional Gaussian potential with N=50 and D=1.0, D=2.0, D=3.0, it can be observed that at a distance of 2, the radial distribution plots converge.
The question that arises here for me is whether I have correctly plotted the graph based on distance. Because based on the shape shown in the article, I expected to see faster convergence (according to the shape, the convergence for N=50 and D=1.0 occurred at a length of 0.75). Additionally, I am aware that my particle number is 20, while in the article, it is 50. However, the desired convergence has not been achieved, and instead, I observe convergence at a length of 5.0. Now, in terms of the "hist_bins" value, I am unsure of the appropriate choice. Should it be equal to L? Can I choose any arbitrary value for the "hist_bins"? In the event that I have encountered fluctuations, what actions should I take to address them?

**Reply by gpescia** [reply]
Hi, 
indeed the pair correlation you compute is not correctly normalized, I just realized sorry. It seems that `density=True` either does not what I thought, or does not work correctly. In the code below I normalize it myself. 
```
N = 20
L = N / 1.0
V = L
sdim = 1

import numpy as np
hist_bins = 80
hist_range = (0, L/2.0)

samples = np.load('../netket/Examples/Continuous/samples.npy').reshape(-1,20)[:2000]
all_histograms = []
for sample in samples:
    d = np.ravel(minimum_distance(sample, 1))
    hist, _ = np.histogram(d, bins=hist_bins, range=hist_range)#, density=True)
    all_histograms.append(hist)

## Computing average histogram##
average_hist = np.mean(all_histograms, axis=0)
bin_edges = np.linspace(hist_range[0], hist_range[1], hist_bins+1)
## Normalize histograms by density##
density = N / (L ** 1)  # Assuming sdim=1
delta_r = bin_edges[1] - bin_edges[0]
normalization = V*delta_r*density
average_hist = average_hist.astype(float)/normalization

plt.plot(average_hist)
plt.xlabel('Distance')
plt.ylabel('Probability Density')
plt.title('Histogram of Distances')
```

**Reply by MrAAafa99** [reply]
Hi Gabriel,
 thank you very much for the corrected code. I also think we should normalize the radial distribution manually. However, it seems that the radial distribution function in the above code is not properly normalized. I realized this when I plotted the code above for Gaussian potentials with 20 particles and densities of 2.0 and 3.0, respectively, and obtained the following shapes:

![download1](https://github.com/netket/netket/assets/54255011/62df7fe7-e224-4201-aa9e-4f36ce53a274)

The above figure depicts the radial distribution plot for a Gaussian potential with 20 particles and a density of 2.0.

![download2](https://github.com/netket/netket/assets/54255011/dfce0253-f960-40a6-8bcc-b88c936ba66e)

The above figure depicts the radial distribution plot for a Gaussian potential with 20 particles and a density of 3.0.

However, when I modify the normalization to be like" normalization = V * delta_r * (density**2) ", ( I cannot determine whether it is correct or not, but this normalization accurately converges the radial distribution plot.) the above plots will converge correctly and yield the following shape:

The plot of a Gaussian potential with 20 particles and a density of 2.0, after correcting the normalization, is as follows:
![download3](https://github.com/netket/netket/assets/54255011/2d779c21-5d08-49b4-8bbb-bf116c47a681)
 
The plot of a Gaussian potential with 20 particles and a density of 3.0, after correcting the normalization, is as follows:
![download 4](https://github.com/netket/netket/assets/54255011/283cc67e-f085-44dc-b992-db1b168a7f83)

However, I still have two remaining questions. Firstly, regarding the presence of " hist_bins", what is the correct value for it in the code? For example, for these two instances (Gaussian potential with densities of 2.0 and 3.0), what is the correct value and what is the reason behind it? Secondly, the more important question is that, as observed, the radial distribution plots still exhibit fluctuations and do not decrease fluctuations even with an increasing number of samples. I aim to plot my own graphs similar to the ones presented in the research paper, which are free from fluctuations. However, despite increasing the number of samples, the fluctuations persist, And my graphs do not resemble the graphs in the paper (shown in the figure below), which is not desirable.
![Screenshot (753)](https://github.com/netket/netket/assets/54255011/22b37fcc-f48c-4f16-a45f-ce1f6f1a73da)

**Reply by gpescia** [reply]
You can choose whatever number of bins you like. If you take many bins you need more samples because the statistics within one bin is reduced. If you take few bins, you don't need many samples but the structure might not be resolved. 
For the fluctuations: You can try to sample multiple times, compute multiple radial distribution functions and average them. Also your energy is better than before but not on top of what I got. It might be necessary to run for longer or take more parameters. How large is the variance on your energy?

**Reply by MrAAafa99** [reply]
In one dimension, I obtained a variance equal to 0.0915 for 20 particles with a Gaussian potential and a density of 2.0. Additionally, for 20 particles with a Gaussian potential and a density of 3.0, the variance is 0.0649. Furthermore, I have calculated the converged energy per particle to be 2.9226 for the example with density 2.0, and the converged energy per particle to be 4.6447 for density 3.0. Both values match the three decimal place accuracy with the numbers obtained from Table 2 in the referenced article.
Regarding the bins, should we not consider bins to be equal to "L=N/D" in terms of physics? For example, for a Gaussian potential with 20 particles and a density of 3.0, when I set bins to 80 according to the graph below,

![download](https://github.com/netket/netket/assets/54255011/03f352ba-9113-499f-9dad-874fc42e2fd6)

 the radial distribution plot converges to 1 at a distance of 40.
 However, for the same example, when I set bins to 150 according to the graph below,
 
![download 2](https://github.com/netket/netket/assets/54255011/2f1668df-b485-46a5-b932-8946141d0181)

the radial distribution plot converges to 1 at a distance of 75, which is ambiguous. Furthermore, I do not know how the horizontal axis (distance) has been dimensionless in the radial distribution plots in the article.

**Reply by gpescia** [reply]
well you plot your histogram against the number of bins not against the distance..
Try `plt.plot(bin_edges[:-1], average_hist)`
And to get a dimensionless distance: `plt.plot(bin_edges[:-1]*density, average_hist)`

**Reply by MrAAafa99** [reply]
Hi Gabriel,
Yes, that's correct. Thank you. However, regarding the fluctuations, as you mentioned, I have converged the energy multiple times and calculated the radial distribution function several times as well. I've taken their averages, but the fluctuations still persist. I still don't know how the clean and smooth curve as presented below has been achieved. 

![Screenshot (753)](https://github.com/netket/netket/assets/54255011/7bb7cc52-abd3-4e26-9444-eba06fc55a63)

Although I had a thought about reducing the fluctuations, which I'm not sure if it's correct or not, after calculating the average of several radial distribution functions, I executed the following code:

```python
# Calculating average histograms.

# Apply moving average to smooth the histogram
window_size = 15
smoothed_hist = np.convolve(average_hist, np.ones(window_size)/window_size, mode='valid')

# Adjust bin_edges to match the size of smoothed_hist
bin_edges = bin_edges[:len(smoothed_hist)]
```
However, I believe there is a physical issue in the code above. And did you also observe fluctuations in the radial distribution function plot provided in the article? If not, what differences did your code have compared to mine that resulted in your plots being smooth and without fluctuations as shown?

**Reply by gcarleo** [reply]
How many samples are you using in total? How many bins do you have?

**Reply by MrAAafa99** [reply]
If by "samples" you mean the number of samples within the VMC, I should say that I usually set the number of samples to 8192. However, I continued increasing the number of samples, and I even tried values like 98,304. As a result, it didn't make a significant difference. But if by "samples" you mean the number of times I minimized the energy, calculated the radial distribution function and took the average, I should mention that I performed this process around 10 times.

Regarding the bins, I generally set them to 100. Nevertheless, I have also experimented with various bin numbers, ranging from 50 to 300. The only consistent observation I've made is that increasing the number of bins leads to more pronounced fluctuations.

**Reply by gcarleo** [reply]
It is important you realize that since you are using a finite number of samples (yes, I mean the VMC samples for the observable you are trying to estimate), there is a statistical uncertainty associated with the histogram you are trying to obtain. Intuitively, the more bins you have and the less samples you have, the larger the uncertainty is. You can have a more quantitative understanding just computing the error bars for yourself. For example, using boostrap: https://chat.openai.com/share/caaa9e43-860a-4638-b16b-33b031883a59 

To give you an idea, if you have 8192 samples from a gaussian and you use 100 bins, these are the kind of error bars you can expect: 

<img width="616" alt="image" src="https://github.com/netket/netket/assets/28149892/37474d12-83d2-46ee-8315-c21fa23da497">

I suspect the fluctuations you are seeing are just due to the error bars, please estimate those.

**Reply by gpescia** [reply]
Hi again, 
I did the optimization myself now for a density of D=3 using your code:
```
import netket as nk
import jax.numpy as jnp

import flax
N = 20
L = N / 3.0

hilb = nk.hilbert.Particle(N=N, L=L, pbc=True)
from optax._src import linear_algebra

def mycb(step, logged_data, driver):
    logged_data["acceptance"] = float(driver.state.sampler_state.acceptance)
    logged_data["globalnorm"] = float(linear_algebra.global_norm(driver._loss_grad))
    return True
    
def minimum_distance(x, sdim):
   """Computes distances between particles using minimum image convention"""
   n_particles = x.shape[0] // sdim
   x = x.reshape(-1, sdim)

   distances = (-x[jnp.newaxis, :, :] + x[:, jnp.newaxis, :])[
       jnp.triu_indices(n_particles, 1)
   ]
   distances = jnp.remainder(distances + L / 2.0, L) - L / 2.0

   return jnp.linalg.norm(distances, axis=1)

sigma = 1 / jnp.sqrt(2)
epsilon = 2.

def potential(x, sdim):
   dis = minimum_distance(x, sdim)
   return jnp.sum(epsilon * (jnp.exp(-(dis)**2)))

ekin = nk.operator.KineticEnergy(hilb, mass=1.0)
pot = nk.operator.PotentialEnergy(hilb, lambda x: potential(x, 1))
ha = ekin + pot

sab = nk.sampler.MetropolisGaussian(hilbert=hilb, sigma=0.4, n_chains=16, n_sweeps=10)
model = nk.models.DeepSetRelDistance(
   hilbert=hilb,
   layers_phi=2,
   layers_rho=3,
   features_phi=(16, 16),
   features_rho=(16, 16, 1),
   cusp_exponent=None
)
vs = nk.vqs.MCState(sab, model, n_samples=1028, n_discard_per_chain=10)
import flax
with open('Gaussian_20_1D_D3_2.mpack', 'rb') as file:
    vs.variables = flax.serialization.from_bytes(vs.variables, file.read())

op = nk.optimizer.Sgd(0.02)
sr = nk.optimizer.SR(diag_shift=0.05)
gs = nk.VMC(ha, op, sab, variational_state=vs, preconditioner=sr)
gs.run(n_iter=100, callback=mycb, out="Gaussian_20_1D_D3")
```
Then I took 1.6 million samples from the state and used the following code to do the radial distribution function (I'll add it to NetKet):
```
import numpy as np
import jax

def minimum_distance(x, sdim, N, L):
    """Computes distances between particles using mimimum image convention"""
    x = x.reshape(-1, N, sdim)
    
    idxs = jnp.triu_indices(N, 1)
    
    distances = (-x[...,jnp.newaxis, :, :] + x[...,:, jnp.newaxis, :])
    distances = distances[...,idxs[0],idxs[1],:]

    distances = jnp.remainder(distances + L / 2.0, L) - L / 2.0

    return jnp.linalg.norm(distances, axis=-1)


def two_body_correlation_g2(samples, n_particles, sdim, L, bins=100):
    dists = minimum_distance(samples, sdim, n_particles, L)
    
    hist, bin_edges = np.histogram(dists.flatten(), bins=bins, range=(0,L/2))    
    delta_r = bin_edges[1] - bin_edges[0]
    centers = (bin_edges[1:] + bin_edges[:-1])/2

    gC = hist / (dists.shape[0]*n_particles)

    def get_jacobian():
        if sdim == 1:
            return np.ones_like(centers) # factor 2 from both sides
        elif sdim == 2:
            return 2*np.pi*centers
        elif sdim == 3:
            return 4*np.pi*centers**2
        else:
            raise NotImplementedError(f"Jacobian factor not yet implemented for sdim={sdim}")
    
    def get_volume():
        return L**sdim
    
    def get_density():
        return n_particles/get_volume()

    gU = get_jacobian()*delta_r*get_density()

    return centers, gC/gU
```
And the resulting radial distribution function looks like this:
![image](https://github.com/netket/netket/assets/79276688/aafab3f0-5722-4287-b11b-473d532c042f)

which is basically free from the fluctuations you see. So I would suggest to take more samples as @gcarleo  pointed out.

**Reply by gcarleo** [reply]
Thank you Gabriel. I guess with the improved estimator you can get the same preicison with much less than million samples, we should look into it when there is time

**Reply by MrAAafa99** [reply]
@gpescia Thank you for sending the corrected code. First and foremost, I must point out that the code you've sent still has an issue.
In the code below, 
`
with open('Gaussian_20_1D_D3_2.mpack', 'rb') as file:
    vs.variables = flax.serialization.from_bytes(vs.variables, file.read())
`
you are attempting to read a file that does not exist, and my guess is that you intended to save the result of the final states. I believe the code should be corrected as follows:
`
with open("Gaussian_20_1D_D3_2.mpack", 'wb') as file:
   file.write(flax.serialization.to_bytes(vs.variables))
`
After correcting this issue, I ran the following code again for myself:
```python
import netket as nk
import jax.numpy as jnp

import flax

N = 20
L = N / 3.0

hilb = nk.hilbert.Particle(N=N, L=L, pbc=True)
from optax._src import linear_algebra


def mycb(step, logged_data, driver):
    logged_data["acceptance"] = float(driver.state.sampler_state.acceptance)
    logged_data["globalnorm"] = float(linear_algebra.global_norm(driver._loss_grad))
    return True


def minimum_distance(x, sdim):
    """Computes distances between particles using minimum image convention"""
    n_particles = x.shape[0] // sdim
    x = x.reshape(-1, sdim)

    distances = (-x[jnp.newaxis, :, :] + x[:, jnp.newaxis, :])[
        jnp.triu_indices(n_particles, 1)
    ]
    distances = jnp.remainder(distances + L / 2.0, L) - L / 2.0

    return jnp.linalg.norm(distances, axis=1)


sigma = 1 / jnp.sqrt(2)
epsilon = 2.


def potential(x, sdim):
    dis = minimum_distance(x, sdim)
    return jnp.sum(epsilon * (jnp.exp(-(dis) ** 2)))


ekin = nk.operator.KineticEnergy(hilb, mass=1.0)
pot = nk.operator.PotentialEnergy(hilb, lambda x: potential(x, 1))
ha = ekin + pot

sab = nk.sampler.MetropolisGaussian(hilbert=hilb, sigma=0.4, n_chains=16, n_sweeps=10)
model = nk.models.DeepSetRelDistance(
    hilbert=hilb,
    layers_phi=2,
    layers_rho=3,
    features_phi=(16, 16),
    features_rho=(16, 16, 1),
    cusp_exponent=None
)
vs = nk.vqs.MCState(sab, model, n_samples=1028, n_discard_per_chain=10)



op = nk.optimizer.Sgd(0.02)
sr = nk.optimizer.SR(diag_shift=0.05)
gs = nk.VMC(ha, op, sab, variational_state=vs, preconditioner=sr)
gs.run(n_iter=100, callback=mycb, out="Gaussian_20_1D_D3")


#write the variational state variable
with open("Gaussian_20_1D_D3_2.mpack", 'wb') as file:
   file.write(flax.serialization.to_bytes(vs.variables))
```
And the optimized energy plot appeared as follows:

![download](https://github.com/netket/netket/assets/54255011/70cae7b0-ce6c-4324-8de0-099815fd45ed)

As can be observed, the converged energy per particle is equal to 4.6444, and the variance is calculated to be 0.0638.
Finally, I loaded the final states using the provided code and calculated the radial distribution according to the code you presented for 1.6 million samples.
```python
# load
import numpy as np
import matplotlib.pyplot as plt
import flax
import netket as nk
import jax.numpy as jnp
with open('Gaussian_20_1D_D3_2.mpack', 'rb') as file:
   vars = flax.serialization.from_bytes(vs.variables, file.read())
vs.variables=vars

x = vs.samples
jnp.save('Gaussian_20_1D_D3_2.npy', x)

# Load samples from file
samples = np.load('Gaussian_20_1D_D3_2.npy').reshape(-1, 20)[:1600000]  # 1.6 million samples


N=20
L = N / 3.0  # Set L according to your requirements
sdim =1

def minimum_distance(x, sdim, N, L):
    """Computes distances between particles using mimimum image convention"""
    x = x.reshape(-1, N, sdim)
    
    idxs = jnp.triu_indices(N, 1)
    
    distances = (-x[...,jnp.newaxis, :, :] + x[...,:, jnp.newaxis, :])
    distances = distances[...,idxs[0],idxs[1],:]

    distances = jnp.remainder(distances + L / 2.0, L) - L / 2.0

    return jnp.linalg.norm(distances, axis=-1)


def two_body_correlation_g2(samples, n_particles, sdim, L, bins=100):
    dists = minimum_distance(samples, sdim, n_particles, L)
    
    hist, bin_edges = np.histogram(dists.flatten(), bins=bins, range=(0,L/2))    
    delta_r = bin_edges[1] - bin_edges[0]
    centers = (bin_edges[1:] + bin_edges[:-1])/2

    gC = hist / (dists.shape[0]*n_particles)

    def get_jacobian():
        if sdim == 1:
            return np.ones_like(centers)  # factor 2 from both sides
        elif sdim == 2:
            return 2*np.pi*centers
        elif sdim == 3:
            return 4*np.pi*centers**2
        else:
            raise NotImplementedError(f"Jacobian factor not yet implemented for sdim={sdim}")
    
    def get_volume():
        return L**sdim
    
    def get_density():
        return n_particles / get_volume()

    gU = get_jacobian() * delta_r * get_density()

    return centers, gC / gU

centers, g2 = two_body_correlation_g2(samples, N, sdim, L)

# Plot the Radial Distribution Function
plt.plot(centers, g2, marker='o')
plt.xlabel('Distance')
plt.ylabel('g2')
plt.title('Radial Distribution Function')
plt.show()
```
And the function g(r) is shown as follows:
![download (1)](https://github.com/netket/netket/assets/54255011/5c312cfd-b238-42c8-8b21-b8e61bf925ce)

As you can observe, even with the increase in the number of samples to 1.6 million, fluctuations persist. This situation persists despite running your unaltered corrected code and having confidence in the convergence outcomes. I've also shared the comprehensive code details in this message to ensure that no part of the code is inadvertently overlooked, resulting in disparate results. However, I still encounter fluctuations that persist.

**Reply by MrAAafa99** [reply]
@gcarleo Thank you, yes, you are absolutely right. However, as you can see from my results, fluctuations persist despite increasing the number of samples (certainly, under these circumstances, it should undoubtedly exhibit lower error bars and consequently, lesser fluctuations) which has left me quite perplexed. I believe I have executed the code correctly unless I've overlooked a detail somewhere.

**Reply by gpescia** [reply, marked as answer]
> you are attempting to read a file that does not exist, and my guess is that you intended to save the result of the final states.

No this part of the code allows to read an .mpack file and overwrite the parameters of the variational state with the parameters from the .mpack file. Like this I can reload my optimized state. You can just comment out this part or replace with the right name of the file. You use it in the second part of the code:
```
with open('Gaussian_20_1D_D3_2.mpack', 'rb') as file:
   vars = flax.serialization.from_bytes(vs.variables, file.read())
vs.variables=vars
```

Are you sure you take 1.6 million samples. In your code I only see:
```
x = vs.samples
jnp.save('Gaussian_20_1D_D3_2.npy', x)
```

which produces 1028 samples according to the code you use to optimize:

`vs = nk.vqs.MCState(sab, model, n_samples=1028, n_discard_per_chain=10)`

Try:
`x = vs.sample(chain_length=10**5)`

**Comment by gcarleo** [comment]
Can we include a g2(r) estimator directly in netket so that everybody can use it?
